core/src/main/java/org/elasticsearch/indices/IndicesModule.java

```diff
@@ -108,14 +128,93 @@ private void registerBuiltinQueryParsers() {
+        registerMapper(IntegerFieldMapper.CONTENT_TYPE, new IntegerFieldMapper.TypeParser());
+        registerMapper(LongFieldMapper.CONTENT_TYPE, new LongFieldMapper.TypeParser());
+        registerMapper(FloatFieldMapper.CONTENT_TYPE, new FloatFieldMapper.TypeParser());
+        registerMapper(DoubleFieldMapper.CONTENT_TYPE, new DoubleFieldMapper.TypeParser());
+        registerMapper(BooleanFieldMapper.CONTENT_TYPE, new BooleanFieldMapper.TypeParser());
+        registerMapper(BinaryFieldMapper.CONTENT_TYPE, new BinaryFieldMapper.TypeParser());
+        registerMapper(DateFieldMapper.CONTENT_TYPE, new DateFieldMapper.TypeParser());
+        registerMapper(IpFieldMapper.CONTENT_TYPE, new IpFieldMapper.TypeParser());
+        registerMapper(StringFieldMapper.CONTENT_TYPE, new StringFieldMapper.TypeParser());
+        registerMapper(TokenCountFieldMapper.CONTENT_TYPE, new TokenCountFieldMapper.TypeParser());
+        registerMapper(ObjectMapper.CONTENT_TYPE, new ObjectMapper.TypeParser());
+        registerMapper(ObjectMapper.NESTED_CONTENT_TYPE, new ObjectMapper.TypeParser());
+        registerMapper(TypeParsers.MULTI_FIELD_CONTENT_TYPE, TypeParsers.multiFieldConverterTypeParser);
+        registerMapper(CompletionFieldMapper.CONTENT_TYPE, new CompletionFieldMapper.TypeParser());
+        registerMapper(GeoPointFieldMapper.CONTENT_TYPE, new GeoPointFieldMapper.TypeParser());
+
+        if (ShapesAvailability.JTS_AVAILABLE) {
+            registerMapper(GeoShapeFieldMapper.CONTENT_TYPE, new GeoShapeFieldMapper.TypeParser());
+        }
+    }
+
+    private void registerBuiltInRootMappers() {
+        // NOTE: the order is important
+
+        // UID first so it will be the first stored field to load (so will benefit from "fields: []" early termination
+        registerRootMapper(UidFieldMapper.NAME, new UidFieldMapper.TypeParser());
+        registerRootMapper(IdFieldMapper.NAME, new IdFieldMapper.TypeParser());
+        registerRootMapper(RoutingFieldMapper.NAME, new RoutingFieldMapper.TypeParser());
+        registerRootMapper(IndexFieldMapper.NAME, new IndexFieldMapper.TypeParser());
+        registerRootMapper(SourceFieldMapper.NAME, new SourceFieldMapper.TypeParser());
+        registerRootMapper(TypeFieldMapper.NAME, new TypeFieldMapper.TypeParser());
+        registerRootMapper(AllFieldMapper.NAME, new AllFieldMapper.TypeParser());
+        registerRootMapper(TimestampFieldMapper.NAME, new TimestampFieldMapper.TypeParser());
+        registerRootMapper(TTLFieldMapper.NAME, new TTLFieldMapper.TypeParser());
+        registerRootMapper(VersionFieldMapper.NAME, new VersionFieldMapper.TypeParser());
+        registerRootMapper(ParentFieldMapper.NAME, new ParentFieldMapper.TypeParser());
+        // _field_names needs to be last in order to see all other fields
+        registerRootMapper(FieldNamesFieldMapper.NAME, new FieldNamesFieldMapper.TypeParser());
```

Comment:
Is it good enough to put this at the end here? Like, if a pugin wanted to add a root mapper.